myia_backend_python/myia_backend_python/python.py

```diff
@@ -0,0 +1,802 @@
+        """
+        graph = self.graph
+        # Register parameters.
+        param_names = []
+        for p in graph.parameters:
+            self.node_to_name[p] = p.debug.debug_name
+            param_names.append(self.node_to_name[p])
+        seq = []
+        # Register nodes.
+        for node in toposort(graph.output, NodeVisitor(), in_graph(graph)):
+            if self.has_node(node):
+                continue
+            elif node.is_constant_graph() and node.value.parent is None:
+                self.node_to_name[node] = self.get_graph_name(node.value)
+            else:
+                self.node_to_name[node] = self.get_new_name()
+                seq.append(node)
+        output = []
+        # Get code for each node.
+        for op in seq:
+            op_name = self.local_ref(op)
+            if op.is_apply():
+                op_code = self.on_apply(op)
+            elif op.is_constant_graph():
+                op_code = self.on_function(op.value, op)
+            elif op.is_constant():
+                op_code = self.on_constant(op)
+            else:
+                raise AssertionError(f"Unsupported node: {op}")
+            # Code for a node may be a list of lines of code.
+            # In such case, latest code line should be only the return value.
+            if isinstance(op_code, list):
+                build_code = op_code[:-1]
+                return_code = op_code[-1]
+                output.extend(build_code)
+                output.append(f"{op_name} = {return_code}")
+            elif op_code:
+                output.append(f"{op_name} = {op_code}")
+        output.append(f"return {self.local_ref(graph.output)}")
+
+        constants_code = []
+        closures_code = []
+        for cst_name, cst_val in self.const_name_to_value.items():
+            constants_code.append(f"{cst_name} = {cst_val}")
+        for fn_name, (fn_params, fn_body) in self.closure_name_to_code.items():
+            fn_signature = f"def {fn_name}({', '.join(fn_params)}):"
+            closures_code.append(fn_signature)
+            closures_code.append(fn_body)
+        # Body code contains constants, then closures, then function code.
+        return param_names, constants_code + closures_code + output
+
+
+class PythonCompiler(_Compiler):
+    """Compile a myia graph into a Python function."""
+
+    def __init__(self):
+        """Initialize."""
+        self.i = 0
+        self._prim_graph_cache = {}
+        self.make_const = PythonConstantConverter()
+        self.graph_to_name = {}
+        self.fn_name_to_code = {}
+
+    def run(self, graph, *args):
+        """Compile given graph."""
+        mng = manage(graph)
+        mng.keep_roots(graph)
+        # Graph to name
+        for g in mng.graphs:
+            if g is graph:
+                self.graph_to_name[g] = "main"
+            else:
+                self.graph_to_name[g] = "_" + g.debug.debug_name
```

Comment:
The names obtained through this method are not very informative. `str(g)` gives a much more informative name, but it contains colons and/or Unicode that's invalid in symbols, and does not guarantee uniqueness. I would suggest adapting `NodeLabeler` in `myia/debug/label.py` so that it can be used to generate valid variable names, and then use that instead of `debug_name`. Then, to guarantee uniqueness, you can have a `collections.Counter` on the side that appends the count for each name to the symbol, unless it's the first one (e.g. we want to generate `x`, `x2`, etc. and `y`, `y2`, ...)

This should also be done for parameters and the `seq` names whenever possible.